vitrina/smart_contracts/forms.py

```diff
@@ -87,19 +97,43 @@ def __init__(self, *args, **kwargs):
         self.helper.form_id = "agreement-upload-form"
         self.helper.add_input(Submit("submit", _("Įkelti dokumentą"), css_class="button is-primary"))
 
-    def clean_file(self) -> FieldFile:
+    def clean_file(self) -> UploadedFile:
         file = self.cleaned_data["file"]
         try:
-            # TODO: check for multiple signatures if AgreementStatuses.INITIATED
-            # TODO: check agreement checksum
-            # TODO: https://github.com/atviriduomenys/katalogas/issues/1706
-            signature_valid = has_valid_signature(file)
-        except InvalidAdocError as error:
-            raise ValidationError(str(error))
-
-        if not signature_valid:
+            with zipfile.ZipFile(file) as zip_file:
+                if not is_valid_adoc(zip_file):
+                    raise InvalidAdocError("Neteisingas ADOC formatas.")
+                if num_of_adoc_root_files(zip_file) > 1:
+                    raise InvalidAdocError(_("Rastas daugiau nei vienas pasirašytas dokumentas."))
+                pdf_path = get_pdf_path_in_adoc(zip_file)
+                if not pdf_path:
+                    raise InvalidAdocError(_("Nerastas PDF dokumentas."))
+                with zip_file.open(pdf_path) as pdf_file:
+                    pdf_bytes = pdf_file.read()
+                if generate_checksum(pdf_bytes) != self.agreement_pdf.checksum:
+                    raise InvalidAdocError(_("PDF dokumentas nesutampa su sutartyje esančiu PDF dokumentu."))
+                signers = get_signers_from_adoc(zip_file)
+        except (InvalidAdocError, zipfile.BadZipFile) as error:
+            raise ValidationError(f"ADOC klaida: {str(error)}")
+
+        num_of_signers = len(signers)
+
+        if num_of_signers == 0:
             raise ValidationError(_("Įkelta sutartis nepasirašyta."))
 
+        if self.agreement.status == AgreementStatuses.FORMED:
+            if num_of_signers > 1:
+                raise ValidationError(_("Įkelta sutartis pasirašyta daugiau nei 1 parašu."))
```

Comment:
More information should be added for the user, e.g.:
```suggestion
                raise ValidationError(_("Įkelta sutartis pasirašyta daugiau nei 1 parašu. Gavėjas turėtų pasirašyti tik vienu parašu."))
```